From 29313e61867f578e2e375e8e2eebe793c0dc1e8e Mon Sep 17 00:00:00 2001
From: Jason <oblitorum@gmail.com>
Date: Tue, 7 Nov 2023 01:16:55 +0800
Subject: [PATCH] Fix pt-visual-explain for 'Impossible ON condition' case

---
 bin/pt-visual-explain | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/bin/pt-visual-explain b/bin/pt-visual-explain
index 53a431c2..f2091e77 100755
--- a/bin/pt-visual-explain
+++ b/bin/pt-visual-explain
@@ -330,7 +330,7 @@ sub transform {
    # Dispatch to a class method to generate the tree.
    # ##################################################################
    my $no_matching_row = join('|',
-      "Impossible (?:WHERE|HAVING)(?: noticed after reading const tables)?",
+      "Impossible (?:WHERE|HAVING|ON)(?: noticed after reading const tables| condition)?",
       'No matching.*row',
       '(?:unique|const) row not found',
    );
@@ -499,6 +499,13 @@ sub ref_or_null {
 
 sub const {
    my ( $self, $row ) = @_;
+   if ( !$row->{key} ) {
+      return {
+         type => 'Constant table access',
+         rows => $row->{rows},
+         children => [$self->table($row)]
+      }
+   }
    return $self->index_access($row, 'Constant index lookup');
 }
 
-- 
2.25.1

